paperqa/clients/openalex.py

```diff
@@ -0,0 +1,233 @@
+from __future__ import annotations
+
+import json
+import logging
+import os
+from collections.abc import Collection
+from datetime import datetime
+from typing import Any
+from urllib.parse import quote
+
+import aiohttp
+
+from paperqa.types import DocDetails
+from paperqa.utils import replace_acute_accents, strings_similarity
+
+from .client_models import DOIOrTitleBasedProvider, DOIQuery, TitleAuthorQuery
+from .exceptions import DOINotFoundError
+from .shared_dicts import BIBTEX_MAPPING
+
+OPENALEX_BASE_URL = "https://api.openalex.org"
+OPENALEX_API_REQUEST_TIMEOUT = 5.0
+
+logger = logging.getLogger(__name__)
+
+
+# author_name will be FamilyName, GivenName Middle initial. (if available)
+# there is no standalone "FamilyName" or "GivenName" fields
+# this manually constructs the name into the format the other clients use
+def reformat_name(name: str) -> str:
+    if "," not in name:
+        return name
+    family, given_names = (x.strip() for x in name.split(",", maxsplit=1))
+    # Return the reformatted name
+    return f"{given_names} {family}"
+
+
+async def get_openalex_mailto() -> str | None:
+    """Get the OpenAlex mailto address.
+
+    Returns:
+        The OpenAlex mailto address if available.
+    """
+    mailto_address = os.environ.get("OPENALEX_MAILTO")
+    if mailto_address is None:
+        logger.warning(
+            "OPENALEX_MAILTO environment variable not set."
+            " your request may be deprioritized by OpenAlex."
+        )
+    return os.environ.get("OPENALEX_MAILTO")
+
+
+async def get_doc_details_from_openalex(
+    session: aiohttp.ClientSession,
+    doi: str | None = None,
+    title: str | None = None,
+    fields: Collection[str] | None = None,
+    title_similarity_threshold: float = 0.75,
+) -> DocDetails | None:
+    """Get paper details from OpenAlex given a DOI or paper title.
+
+    Args:
+        session: The active session of the request.
+        doi: The DOI of the paper.
+        title: The title of the paper.
+        fields: Specific fields to include in the request.
+        title_similarity_threshold: The threshold for title similarity.
+
+    Returns:
+        The details of the document if found, otherwise None.
+
+    Raises:
+        ValueError: If neither DOI nor title is provided.
+        DOINotFoundError: If the paper cannot be found.
+    """
+    mailto = await get_openalex_mailto()
+    params = {"mailto": mailto} if mailto else {}
+
+    if doi is title is None:
+        raise ValueError("Either a DOI or title must be provided.")
+
+    url = f"{OPENALEX_BASE_URL}/works"
+    if doi:
+        url += f"/https://doi.org/{quote(doi, safe='')}"
```

Comment:
Does this make sense? `f"{OPENALEX_BASE_URL}/works/https://doi.org/{quote(doi, safe='')}"`

This seems like a mistake to me

Reply:
it's not, it's weird.

https://api.openalex.org/works/https://doi.org/10.7717/peerj.4375

is valid, for example

Reply:
might be worth a comment so someone doesn't accidentally undo this actually. 
I also thought it was strange.

Reply:
Oh wow that is interesting. Yes, adding a comment sharing that example would be good